File: backend/web/src/main/java/net/dorokhov/pony/web/domain/UserDto.java
package net.dorokhov.pony.web.domain;

import com.google.common.collect.ImmutableSet;
import net.dorokhov.pony.api.user.domain.User;

import javax.annotation.Nullable;
import java.time.LocalDateTime;
import java.util.Set;

import static com.google.common.base.Preconditions.checkNotNull;

public final class UserDto extends BaseDto {

    public enum Role {

        USER, ADMIN;

        public Set<User.Role> convert() {
            if (this == ADMIN) {
                return ImmutableSet.of(User.Role.USER, User.Role.ADMIN);
            } else {
                return ImmutableSet.of(User.Role.USER);
            }
        }

        public static Role of(Set<User.Role> roles) {
            if (roles.contains(User.Role.ADMIN)) {
                return Role.ADMIN;
            } else {
                return Role.USER;
            }
        }
    }

    private final String name;
    private final String email;

    private final Role role;

    UserDto(Long id, LocalDateTime creationDate, @Nullable LocalDateTime updateDate,
            String name, String email, Role role) {
        super(id, creationDate, updateDate);
        this.name = checkNotNull(name);
        this.email = checkNotNull(email);
        this.role = checkNotNull(role);
    }

    public String getName() {
        return name;
    }

    public String getEmail() {
        return email;
    }

    public Role getRole() {
        return role;
    }

    public static UserDto of(User user) {
        return new UserDto(user.getId(), user.getCreationDate(), user.getUpdateDate(),
                user.getName(), user.getEmail(), Role.of(user.getRoles()));
    }
}

File: frontend/src/app/domain/user.dto.ts
import {PageDto} from "./common.dto";

export interface UserDto {
  id: string;
  creationDate: string;
  updateDate: string | undefined;
  name: string;
  email: string;
  role: UserDto.Role;
}

export namespace UserDto {
  export enum Role {
    USER = 'USER',
    ADMIN = 'ADMIN',
  }
}

export interface AuthenticationDto {
  user: UserDto;
  accessToken: string;
  staticToken: string;
}

export interface UserPageDto extends PageDto {
  users: UserDto[];
}

export interface UserUpdateCommandDto {
  id: string;
  name: string;
  email: string;
  newPassword: string;
  role: UserDto.Role;
}

export interface UserCreateCommandDto {
  name: string;
  email: string;
  password: string;
  role: UserDto.Role;
}

export interface CurrentUserUpdateCommandDto {
  name: string;
  email: string;
  oldPassword: string;
  newPassword: string;
}
